bw_check_commit_hook/__init__.py

```diff
@@ -0,0 +1,205 @@
+#!/usr/bin/env python3
+# Copyright (c) Blu Wireless Technology Ltd. 2023, All Rights Reserved
+
+import argparse
+import os
+import sys
+from collections import defaultdict
+from datetime import datetime
+from pathlib import Path
+from subprocess import PIPE, check_output, run
+from typing import DefaultDict, Dict, List, Optional, Union
+from pkg_resources import resource_filename
+import yaml
+
+# template used to add a patch-style header to a diff
+PATCH_TEMPLATE = """\
+From: A Non <a.non@bluwireless.com>
+Date: {current_datetime}
+Subject: [PATCH] {commit_msg}
+---
+{diff}
+--
+"""
+
+def _parse_args() -> Dict[str, Union[Path, List[Path]]]:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--ci",
+        action="store_true",
+        help="Indicates that the (manual) hook is being run in CI",
+    )
+    parser.add_argument(
+        "files",
+        type=Path,
+        nargs="*",
+        default=[],
+        help="The relevant files",
+    )
+    default_config_file = Path(resource_filename(__name__, "data/checkpatch.yaml"))
+    parser.add_argument(
+        "--config-file",
+        type=Path,
+        default=default_config_file,
+        help="Path to config file",
+    )
+    return vars(parser.parse_args())
+
+
+def _load_config(config_file: Path) -> Optional[Dict]:
+    mandatory_keys = ["DIR_CONFIGS"]
+    try:
+        with config_file.open("r") as file:
+            config = yaml.safe_load(file)
+        for key in mandatory_keys:
+            if key not in config:
+                print(f"Missing mandatory key {key} in config file")
+                return None
+        return config
+    except Exception as e:
+        print(f"Error loading YAML config file: {e}")
+        return None
+
+
+def pre_commit_hook(running_in_ci: bool, commit_files: List[Path], config_file: Path) -> None:
+    config = _load_config(config_file)
+    if config is None:
+        sys.exit(1)
+
+    if running_in_ci:
+        commit_files = _get_commit_files_from_log()
+
+    errors = defaultdict(lambda: [])
+    # accumulate checkpatch errors for each configured dir
+    for config_dir, config in config["DIR_CONFIGS"].items():
+        # Default values
+        config['errors_enabled'] = config.get('errors_enabled', [])
+        config['errors_ignored'] = config.get('errors_ignored', [])
+        config['max_line_length'] = config.get('max_line_length', 120)
+
+        # Merge with defaults
+        config['errors_enabled'].extend(config.get("ERRORS_ENABLED", []))
+        config['errors_ignored'].extend(config.get("ERRORS_IGNORED", []))
```

Comment:
Could this determination of what config to apply within a directory be refactored into a separate function?

This code also just seems a bit strange, for example it is relying on very similar key names where one is uppercase and the other lowercase.